### PR TITLE
feat(Clarity2) Get burnchain_header_hash by burnchain_height

### PR DESCRIPTION
**Pull request by gregorycoppola**
Fixes https://github.com/blockstack/stacks-blockchain/issues/2663

## Description

This change allows a clarity contract to get the **header hash** of a **burnchain block** using the ***burchain* height**. Currently, you can get the header hash of the burnchain only using the ***Stacks* height**.

## Testing

Tests have been added in the following files:
```
M	src/chainstate/stacks/boot/contract_tests.rs
M	src/vm/analysis/arithmetic_checker/tests.rs
M	src/vm/analysis/type_checker/tests/mod.rs
M	src/vm/tests/contracts.rs
M	src/vm/tests/costs.rs
```

A TODO was added to modify the following integration test. We avoid doing this now because it will require having a way to configure the Stacks Epoch for the Mocknet miner.
```
M	testnet/stacks-node/src/tests/integrations.rs
```

### CODE REVIEW

File: src/vm/database/clarity_db.rs
Comment (gregorycoppola) [inline src/vm/database/clarity_db.rs:715]: Note: New implementation.

File: src/clarity_vm/database/mod.rs
Comment (gregorycoppola) [inline src/clarity_vm/database/mod.rs:34]: New implementation

**Review by gregorycoppola — COMMENTED** [review]
added some pointers to implementations

File: src/vm/database/clarity_db.rs
Comment (kantai) [inline src/vm/database/clarity_db.rs:731]: This looks like the right approach to me!

File: src/clarity_vm/database/mod.rs
Comment (kantai) [inline src/clarity_vm/database/mod.rs:179]: I think this is right approach. You should be more careful about the error handling here. Basically, we want the "burn header hash" to:

* Return `Some(burn_header_hash)` if the given height is `0 <= height < current_burn_block_height`
* Return `None` if the given height is `>= current_burn_block_height`

Anything else should _panic_ (including errors from the SortitionDB: any such error means the database is corrupt, and the node should not be trying to continue processing Stacks transactions).

File: testnet/stacks-node/src/tests/integrations.rs
Comment (gregorycoppola) [inline testnet/stacks-node/src/tests/integrations.rs:371]: this is the test so far

File: src/vm/docs/mod.rs
Comment (pavitthrap) [inline src/vm/docs/mod.rs:1346]: I think you should clarify that the block height input might refer to either the burn chain or the. stacks blockchain, depending on the property

**Review by pavitthrap — COMMENTED** [review]
You should add tests to type checker and arithmetic checker, as well as `vm/tests/contracts.rs`. Trying seeing where else `get-block-info?` is tested.

File: src/vm/docs/mod.rs
Comment (gregorycoppola) [reply src/vm/docs/mod.rs:1346]: Aaron suggested making a new function called "get-burn-block-info?" instead.

Do you think it would be easier to make the docs if we did that?

https://github.com/blockstack/stacks-blockchain/issues/2663#issuecomment-866154516

File: src/vm/database/clarity_db.rs
Comment (pavitthrap) [inline src/vm/database/clarity_db.rs:109]: should this function be named something more like `get_burn_header_hash_using_height` instead?

File: src/vm/docs/mod.rs
Comment (pavitthrap) [reply src/vm/docs/mod.rs:1346]: Oh - yes, I definitely prefer the approach of putting this functionality into a separate clarity function, I think that is way less prone to user error

**Review by pavitthrap — COMMENTED** [review]
Agree with Aaron's idea of putting this functionality into `get-burn-block-info?`

Also, need to add type checker & arithmetic checker tests still I think

File: src/vm/docs/mod.rs
Comment (friedger) [reply src/vm/docs/mod.rs:1346]: Yes. In the future `get-burn-block-info?` could also take `stacks-tip` and return the stacks tip.

File: src/vm/database/clarity_db.rs
Comment (pavitthrap) [inline src/vm/database/clarity_db.rs:111]: Do you want to add a markdown section for "Panics"?

File: src/vm/database/clarity_db.rs
Comment (pavitthrap) [reply src/vm/database/clarity_db.rs:127]: what were your thoughts here?

File: src/vm/docs/mod.rs
Comment (pavitthrap) [inline src/vm/docs/mod.rs:1359]: nit: returns *the* header hash

File: src/vm/docs/mod.rs
Comment (pavitthrap) [inline src/vm/docs/mod.rs:1372]: Should we name the second arg `BlockHeightUint` instead?

File: src/vm/functions/database.rs
Comment (pavitthrap) [inline src/vm/functions/database.rs:528]: Do we want to be returning an error here? Returning Value::none() is not too informative of what went wrong?

File: src/vm/tests/contracts.rs
Comment (pavitthrap) [inline src/vm/tests/contracts.rs:175]: Could we use `TestSimBurnStateDB` here?

File: src/vm/functions/database.rs
Comment (gregorycoppola) [reply src/vm/functions/database.rs:531]: This is copy-pasted.. I guess the assumption was that we would never overflow u32 in practice. I.e., if `result` is u128, and we can't cast to u32, then it is bigger than u32::MAX.

Can we assume that?

File: src/vm/docs/mod.rs
Comment (gregorycoppola) [reply src/vm/docs/mod.rs:1372]: I copy/pasted this from "get-block-info".

You could argue that the "Height" implies that it is unsigned and so "BlockHeightUInt" is redundant.

File: src/vm/docs/mod.rs
Comment (gregorycoppola) [reply src/vm/docs/mod.rs]: done

File: src/vm/database/clarity_db.rs
Comment (gregorycoppola) [reply src/vm/database/clarity_db.rs:111]: Oh yeah.. if that's our convention I should.. you know it's going to add a lot of vertical height to the comments if we do this. The docs also suggest a special section for "#Errors":
https://rust-lang.github.io/api-guidelines/documentation.html

File: src/vm/database/clarity_db.rs
Comment (gregorycoppola) [reply src/vm/database/clarity_db.rs:127]: oh whoops.. Well, the reason it is "using_canonical_sortition" is because the "canonical sortiiton" is the hidden argument. That is, the `height` part is in the signature. Compare with `get_burn_header_hash`. The idea is that these two functions are similar but this one uses an implicit sortition. `get_burn_header_hash` also doesn't mention height.

File: src/vm/docs/mod.rs
Comment (gregorycoppola) [reply src/vm/docs/mod.rs]: Did this.. turned out to be a big delta but worth it!

File: src/clarity_vm/database/mod.rs
Comment (gregorycoppola) [reply src/clarity_vm/database/mod.rs]: did this

**Review by gregorycoppola — COMMENTED** [review]
@pavitthrap thanks for the comments.

File: src/vm/tests/contracts.rs
Comment (pavitthrap) [reply src/vm/tests/contracts.rs:175]: I mean I guess I'm wondering if another mock of `BurnStateDB` is strictly necessary - if it seems necessary to you, then go ahead with `BurnBlockTestDB`.

File: src/vm/docs/mod.rs
Comment (pavitthrap) [reply src/vm/docs/mod.rs:1372]: I see your point. 

Looking more broadly at this file though, it seems that `input_type` often just uses the actual clarity types, leaving the more descriptive names to the `signature` field. I think I would prefer that convention here.

File: src/vm/database/clarity_db.rs
Comment (pavitthrap) [reply src/vm/database/clarity_db.rs:111]: Ya, I think that makes sense to me, since that what the convention would be with rustdoc.

File: src/vm/functions/database.rs
Comment (pavitthrap) [reply src/vm/functions/database.rs:531]: Right, but this branch represents the error case of overflow, so I feel like the result should be a corresponding error.

File: src/vm/database/clarity_db.rs
Comment (pavitthrap) [reply src/vm/database/clarity_db.rs:127]: ahh I see, sounds good to me.

**Review by pavitthrap — COMMENTED** [review]
Looks good, just awaiting your replies on the ongoing comment threads we have.

File: src/vm/database/clarity_db.rs
Comment (gregorycoppola) [reply src/vm/database/clarity_db.rs:112]: ok.. can start doing this

File: src/vm/docs/mod.rs
Comment (gregorycoppola) [reply src/vm/docs/mod.rs:1372]: I was actually copying `get-block-info`.. should I just change both?

File: src/vm/functions/database.rs
Comment (gregorycoppola) [reply src/vm/functions/database.rs:531]: This would have to be a user facing error? which one would it be? https://github.com/blockstack/stacks-blockchain/blob/4650aae6ffb111d487ce324fd8591c529e60c24e/src/vm/analysis/errors.rs#L27

File: src/vm/tests/contracts.rs
Comment (gregorycoppola) [reply src/vm/tests/contracts.rs:175]: it's necessary.. we need a mock otherwise we would have to spin up a whole database.. which is a good integration test, but wouldn't be a unit test.

also, we can't use NullBurnStateDB because it always returns None and then we can't do a meaningufl test

**Review by gregorycoppola — COMMENTED** [review]
@pavitthrap  I made the "#Panics" changes.. had a few questions.

File: src/vm/docs/mod.rs
Comment (pavitthrap) [reply src/vm/docs/mod.rs:1372]: I think that would be good, it would make the documentation more consistent overall

File: src/vm/functions/database.rs
Comment (pavitthrap) [reply src/vm/functions/database.rs:531]: I think `CheckErrors::ValueTooLarge` makes sense here

File: src/vm/docs/mod.rs
Comment (gregorycoppola) [reply src/vm/docs/mod.rs]: done

**Review by gregorycoppola — COMMENTED** [review]
@pavitthrap just the "None" vs "Error" issue remaining

File: src/vm/functions/database.rs
Comment (pavitthrap) [reply src/vm/functions/database.rs:531]: You're right, that error type doesn't make sense. 

I'm alright with returning `Value::none()` here.

**Review by pavitthrap — APPROVED** [review]
LGTM! Thanks for addressing all the feedback :)

File: src/vm/functions/database.rs
Comment (jcnelson) [inline src/vm/functions/database.rs:506]: comment typo -- did you mean "`get-burn-block-info?`"?

File: src/vm/analysis/type_checker/natives/mod.rs
Comment (jcnelson) [inline src/vm/analysis/type_checker/natives/mod.rs:533]: Up to you how you decide to do this, but do you want a separate `CheckErrors` error for burn block properties not being known to the system?

File: src/vm/database/clarity_db.rs
Comment (kantai) [inline src/vm/database/clarity_db.rs:735]: I don't think that it is sufficient to use the canonical sortition for looking up the burnchain block at a given burnchain height here. The way that these lookups are performed is that the provided "sortition" defines the burnchain fork that the lookup is performed from. The canonical sortition _may not_ be the same fork that the currently-being-evaluated Stacks block was contained within (I believe that with the way the ChainsCoordinator currently works, this _is_ the case, but it's not a guarantee that the ChainsCoordinator explicitly provides).

For example, the burnchain could look like this:

```
A -> B -> C -> D
 \_> B' -> C'
```

Where the currently-being-evaluated Stacks block was elected in `C'`, but the canonical sortition is `D`. We'd want `get-burn-block-info u1` to return `B'` in this case and _not_ `B`.

The second important thing here is that the height supplied to `get-burn-block-info` _must_ be strictly less than the current burn block height. This is because miners have to evaluate the block _before_ the burn block is mined, so they could not _know_ that `(get-burn-block-info u2)` should return `C'`. Basically, if the supplied height is `>= current_burn_block_height` it should return `none`.

Comment (jcnelson) [reply src/vm/database/clarity_db.rs:735]: Beat me to it.

You're going to need to track the `StacksBlockId` of the chain tip at which this code is running, and use *that* to look up the associated burnchain block hash.  Then, if the VM is in the process of building block `C'` off of `B'`, then this method would be able to report the hash of `B'` as the hash of the block at height 2.  Correspondingly, if the VM is in the process of building block `C` off of `B`, then this method would be able to report the hash of `B` as the hash of the block at height 2.

File: src/vm/database/clarity_db.rs
Comment (jcnelson) [inline src/vm/database/clarity_db.rs:132]: You should probably just drop this method altogether.  This method is going to need to take the `StacksBlockId` of the chain tip that the VM is currently working on, in addition to the burnchain block height.  It would look something like this:

```
// just an example; feel free to rename
fn get_burn_header_hash_from_tip(&self, stacks_tip: &StacksBlockId, burn_height: u32) -> Option<BurnchainHeaderHash>)
```

File: src/vm/database/clarity_db.rs
Comment (gregorycoppola) [inline src/vm/database/clarity_db.rs:727]: @kantai said "Basically, if the `supplied height is >= current_burn_block_height` it should return none."

I conceptually get why this is, I think, because we are "building the block" and the snapshot wouldn't exist in sqlite3 yet.

But, what is the literal API call for `current_burn_block_height`.

File: src/vm/database/clarity_db.rs
Comment (gregorycoppola) [inline src/vm/database/clarity_db.rs:735]: Should these None cases 1) propagate None, 2) return Error, 3) panic?

File: src/vm/database/clarity_db.rs
Comment (gregorycoppola) [inline src/vm/database/clarity_db.rs:727]: @kantai @jcnelson can you review the logic in this method before I re-write the tests?

**Review by gregorycoppola — COMMENTED** [review]
@kantai @jcnelson I'm just asking for an *intermediate* review of the logic of `ClarityDatabase::get_burnchain_block_header_hash_for_burnchain_height` before I re-write the tests and clean up the PR.

The reset of the PR doesn't need/isn't ready for review.

**Comment by jcnelson** [comment]
Blocked on getting a mini-LGTM on the logic for obtaining a burnchain header hash off of a particular chain tip. CC @jcnelson @kantai

File: src/vm/database/clarity_db.rs
Comment (kantai) [inline src/vm/database/clarity_db.rs:728]: Is this the current Stack height or the current burn height? Each Stacks block has both (and they aren't equal), i.e., a Stacks block could be Stacks height 100, but burn height 667000.

File: src/vm/database/clarity_db.rs
Comment (kantai) [inline src/vm/database/clarity_db.rs:729]: You should use the _parent_ ID block hash, rather than the current stacks height. The reason for that is that the current block hash is evaluated with a placeholder by the miner, so its consensus hash, etc. isn't contained in the headers db yet (which only contains information about already-mined blocks).

File: src/vm/database/clarity_db.rs
Comment (kantai) [reply src/vm/database/clarity_db.rs:735]: It should panic. _Except_ in the case that the current block height is 0. That is, a parent block should _always_ have a consensus hash (unless it's at height 0 and there is no parent).

File: src/vm/database/clarity_db.rs
Comment (kantai) [inline src/vm/database/clarity_db.rs:740]: Panic

File: src/vm/database/clarity_db.rs
Comment (kantai) [inline src/vm/database/clarity_db.rs:723]: I believe this returns `Some() if first_burn_block_height <= height < current_burn_block_height` and None otherwise. The sortition DB _does not_ track information about burn blocks before `first_burn_block_height`.

File: src/vm/database/clarity_db.rs
Comment (kantai) [inline src/vm/database/clarity_db.rs:751]: As part of this PR, can you add a rustdoc to the interface method `get_burn_header_hash` that indicates that it returns the burnchain header has for the given burn block height, as queried from the given Sortition ID?

File: src/chainstate/burn/db/sortdb.rs
Comment (jcnelson) [inline src/chainstate/burn/db/sortdb.rs:3708]: "snapshots"

Yes, this happens once we have processed and stored the bitcoin block and extracted and processed all of its stacks-specific transactions.  There is one snapshot per sortition, which is a view calculated over the set of block-commits in a bitcoin block.

File: src/clarity_vm/database/mod.rs
Comment (jcnelson) [inline src/clarity_vm/database/mod.rs:167]: How is `height < 0` if it's a `u32`?

File: src/vm/database/clarity_db.rs
Comment (jcnelson) [reply src/vm/database/clarity_db.rs:727]: This code does not work in the presence of Bitcoin forks.  The sortition DB and burnchain DB will process and store state for multiple different Bitcoin blocks at the same height, should a Bitcoin reorg happen.

If you want to query a burnchain block hash at a particular height, you need to know which burnchain fork you're on.  You can do this given the Stacks chain tip's ID block hash -- i.e. the ID block hash of the parent block that this Clarity code is building on top of.

The `id_bhh` variable you load below should be the _same ID hash_ that was passed to the inner `MarfedKV`'s `begin()` method as `current`.  This, in turn, is the ID hash passed as `current` to `ClarityInstance::begin_block()`.  You should *not* be trying to deduce `id_bhh` from a block height.

File: src/vm/database/clarity_db.rs
Comment (jcnelson) [reply src/vm/database/clarity_db.rs:729]: Per my other comment, this would be the same as the value of `current` passed into `ClarityInstance::begin_block()`.

File: src/vm/docs/mod.rs
Comment (jcnelson) [inline src/vm/docs/mod.rs:1384]: I'm not sure `u0` is a good example here.

File: src/vm/docs/mod.rs
Comment (jcnelson) [inline src/vm/docs/mod.rs:1377]: It will also return `none` for any burnchain block value prior to the start of the Stacks chain (i.e. 666050).  We can discuss whether or not this is desirable, but that's how the code currently behaves.

File: testnet/stacks-node/src/tests/integrations.rs
Comment (jcnelson) [inline testnet/stacks-node/src/tests/integrations.rs:53]: Specific to this test, can you verify that `burn-block-info` will continue to work even if there was no Stacks block chosen at that height?  Also, verify that we get `(none)` for any block height prior to the first burnchain block height.

**Review by gregorycoppola — COMMENTED** [review]
OK. I implemented the algo changes here

**Comment by hstove** [comment]
Oh wow this is so handy, thanks! Very critical for validating burnchain transactions.

**Comment by hstove** [comment]
Question on this - if the burnchain block at that height does not correspond to a stacks block (i.e. because of a flash block), will this return `none`? From reading the code, it seems like it would return `none`, but I'm not sure.

**Comment by jcnelson** [comment]
I'm going to take over working on this.

**Comment by jcnelson** [comment]
This is now ready for review @kantai @gregorycoppola @lgalabru @obycode

File: clarity/src/vm/functions/database.rs
Comment (obycode) [inline clarity/src/vm/functions/database.rs:779]:
```suggestion
/// - CheckErrors::NoSuchBurnBlockInfoProperty if `args[0]` isn't a BurnBlockInfoProperty.
/// - CheckErrors::TypeValueError if `args[1]` isn't a uint
```

File: clarity/src/vm/analysis/type_checker/natives/mod.rs
Comment (jcnelson) [reply clarity/src/vm/analysis/type_checker/natives/mod.rs:535]: So, `special_get_burn_block_info()` uses `check_argument_count()`, meaning that passing the wrong number of arguments would be a runtime error, not an analysis error.  Is this desired @gregorycoppola @kantai?

Comment (jcnelson) [reply clarity/src/vm/analysis/type_checker/natives/mod.rs:535]: To be clear, we can totally change it for those functions too as part of 2.1; we just can't do it today ;)  Happy to use `check_argument_count()` here.

File: clarity/src/vm/analysis/type_checker/tests/mod.rs
Comment (jcnelson) [reply clarity/src/vm/analysis/type_checker/tests/mod.rs]: Removed

File: clarity/src/vm/analysis/type_checker/natives/mod.rs
Comment (jcnelson) [reply clarity/src/vm/analysis/type_checker/natives/mod.rs]: Yup, agreed.  Using `check_argument_count()` here.  Also: https://github.com/stacks-network/stacks-blockchain/issues/3117

File: clarity/src/vm/analysis/type_checker/natives/mod.rs
Comment (jcnelson) [reply clarity/src/vm/analysis/type_checker/natives/mod.rs]: Done

File: clarity/src/vm/docs/mod.rs
Comment (jcnelson) [reply clarity/src/vm/docs/mod.rs]: Yup, fixed

File: clarity/src/vm/functions/database.rs
Comment (jcnelson) [reply clarity/src/vm/functions/database.rs:779]: Fixed

**Review by obycode — APPROVED** [review]

**Pull request merged**
